src/custom-client-scripts/client-script.js

```diff
@@ -0,0 +1,99 @@
+import { readFile } from '../utils/promisified-functions';
+import { GeneralError } from '../errors/runtime';
+import { RUNTIME_ERRORS } from '../errors/types';
+import resolvePathRelativelyCwd from '../utils/resolve-path-relatively-cwd';
+import { RequestFilterRule, generateUniqueId } from 'testcafe-hammerhead';
+
+const BEAUTIFY_REGEXP = /[/.:\s\\]/g;
+const BEAUTIFY_CHAR   = '_';
+
+const EMPTY_CONTENT_STR      = '{ content: <empty> }';
+const CONTENT_STR_MAX_LENGTH = 30;
+const CONTENT_ELLIPSIS_STR   = '...';
+
+const URL_UNIQUE_PART_LENGTH = 7;
+
+export default class ClientScript {
+    constructor (init) {
+        this.init    = init || null;
+        this.url     = generateUniqueId(URL_UNIQUE_PART_LENGTH);
+        this.content = '';
+        this.path    = null;
+        this.module  = null;
+        this.page    = RequestFilterRule.ANY;
+    }
+
+    async _loadFromPath (path) {
+        try {
+            const resolvedPath = resolvePathRelativelyCwd(path);
+
+            this.path    = resolvedPath;
+            this.content = await readFile(this.path);
+            this.content = this.content.toString();
+            this.url     = path || this.url;
+        }
+        catch (e) {
+            throw new GeneralError(RUNTIME_ERRORS.cannotLoadClientScriptFromPath, path);
+        }
+    }
+
+    async _loadFromModule (name) {
+        const resolvedPath = require.resolve(name);
+
+        await this._loadFromPath(resolvedPath);
+    }
+
+    _prepareUrl () {
+        this.url = this.url.replace(BEAUTIFY_REGEXP, BEAUTIFY_CHAR).toLowerCase();
+    }
+
+    async load () {
+        if (this.init === null)
+            throw new GeneralError(RUNTIME_ERRORS.clientScriptInitializerIsNotSpecified);
+        else if (typeof this.init === 'string')
+            await this._loadFromPath(this.init);
+        else {
+            const { path: initPath, content: initContent, module: initModule, page: initPage } = this.init;
+
+            if (initPath && initContent || initPath && initModule || initContent && initModule)
+                throw new GeneralError(RUNTIME_ERRORS.clientScriptInitializierMultipleContentSources);
+
+            if (initPath)
+                await this._loadFromPath(initPath);
+            else if (initModule)
+                await this._loadFromModule(initModule);
```

Comment:
What if the specified module is not in the UMD format and can't be loaded as a single file, because it imports other files/modules?

Reply:
This case is not supported.
We don't discuss this scenario, our customers don't ask us too. This is why, I propose to not implement that until first customer request.

Reply:
I don't ask about implementing this scenario. We should make a clear error message to the user if it tries to load a module in an incorrect format.

Reply:
Ok. I will do it.